src/rgw/rgw_reshard.cc

```diff
@@ -285,174 +272,303 @@ int RGWBucketReshard::set_resharding_status(const DoutPrefixProvider *dpp,
-                                                 const DoutPrefixProvider *dpp)
-{
-  return ::create_new_bucket_instance(store, new_num_shards,
-				      bucket_info, bucket_attrs, new_bucket_info, dpp);
-}
+  if (ret = fault.check("block_writes");
+      ret == 0) { // no fault injected, block writes to the current index shards
+    ret = set_resharding_status(dpp, static_cast<rgw::sal::RadosStore*>(store), bucket_info,
+                                cls_rgw_reshard_status::IN_PROGRESS);
+  }
 
-int RGWBucketReshard::cancel(const DoutPrefixProvider *dpp)
-{
-  int ret = reshard_lock.lock(dpp);
   if (ret < 0) {
+    ldpp_dout(dpp, 0) << "ERROR: " << __func__ << " failed to pause "
+        "writes to the current index: " << cpp_strerror(ret) << dendl;
+    // clean up the target layout (ignore errors)
+    revert_target_layout(store, bucket_info, fault, dpp);
     return ret;
   }
+  return 0;
+} // init_reshard
 
-  ret = clear_resharding(dpp);
+static int cancel_reshard(rgw::sal::RadosStore* store,
+                          RGWBucketInfo& bucket_info,
+                          const ReshardFaultInjector& fault,
+                          const DoutPrefixProvider *dpp)
+{
+  // unblock writes to the current index shard objects
+  int ret = set_resharding_status(dpp, static_cast<rgw::sal::RadosStore*>(store), bucket_info,
+                                  cls_rgw_reshard_status::NOT_RESHARDING);
+  if (ret < 0) {
+    ldpp_dout(dpp, 1) << "WARNING: " << __func__ << " failed to unblock "
+        "writes to current index objects: " << cpp_strerror(ret) << dendl;
+    ret = 0; // non-fatal error
+  }
 
-  reshard_lock.unlock();
-  return ret;
-}
+  if (bucket_info.layout.target_index) {
+    return revert_target_layout(store, bucket_info, fault, dpp);
+  }
+  // there is nothing to revert
+  return 0;
+} // cancel_reshard
 
-class BucketInfoReshardUpdate
+static int commit_reshard(rgw::sal::RadosStore* store,
+                          RGWBucketInfo& bucket_info,
+                          const ReshardFaultInjector& fault,
+                          const DoutPrefixProvider *dpp)
 {
-  const DoutPrefixProvider *dpp;
-  rgw::sal::RadosStore* store;
-  RGWBucketInfo& bucket_info;
-  std::map<string, bufferlist> bucket_attrs;
+  auto& layout = bucket_info.layout;
+  auto prev = layout; // make a copy for cleanup
+  const auto next_log_gen = layout.logs.back().gen + 1;
 
-  bool in_progress{false};
+  bool remove_index = true;
 
-  int set_status(cls_rgw_reshard_status s, const DoutPrefixProvider *dpp) {
-    bucket_info.reshard_status = s;
-    int ret = store->getRados()->put_bucket_instance_info(bucket_info, false, real_time(), &bucket_attrs, dpp);
-    if (ret < 0) {
-      ldpp_dout(dpp, 0) << "ERROR: failed to write bucket info, ret=" << ret << dendl;
-      return ret;
+  if (!store->svc()->zone->need_to_log_data()) {
+    // if we're not syncing data, we can drop any existing logs
+    layout.logs.clear();
```

Comment:
@cbodley Is this safe? We use layout.logs.back() in enough places that wiping out all logs doesn't seem like something we can get away with.

Reply:
accessing logs.back() outside of multisite code does not seem safe, no

i think this reshard code is correct, because outside of multisite, we'll always delete the previous generation's index objects. so logs.back() would point at missing objects if we left it in the vector

as @ivancich found in https://github.com/ceph/ceph/pull/44393, indexless buckets are another case where we wouldn't have any logs

Reply:
er, this is in `commit_reshard()`, which goes on to add the *new* index layout to `logs` with:
```c++
layout.logs.push_back(log_layout_from_index(next_log_gen, layout.current_index));
```
so even with multisite disabled, we'll still be able to rely on `logs.back()`